### ISSUE TITLE
Intermittent build failures on 24.04 related to compiling dynamic models

### ISSUE THREAD

**Issue by fcooper8472** [Chaste/Chaste#343]
See e.g. this test failure during compilation:

https://github.com/Chaste/Chaste/actions/runs/11936239185/job/33303134920

```bash
[  2%] Linking CXX shared library dynamic/libDynamicallyLoadableLr91.so
/usr/bin/ld: cannot find CMakeFiles/DynamicallyLoadableLr91.dir/link.d: No such file or directory
collect2: error: ld returned 1 exit status
gmake[3]: *** [heart/CMakeFiles/DynamicallyLoadableLr91.dir/build.make:101: heart/dynamic/libDynamicallyLoadableLr91.so] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:20999: heart/CMakeFiles/DynamicallyLoadableLr91.dir/all] Error 2
```

**Comment by fcooper8472** [comment]
This appears to relate to the compilation of dynamic cell models. These are compiled completely separate to Chaste, but therefore contain Chaste symbols that will be undefined when the models themselves are linked.

These symbols are intended to be looked up dynamically at runtime, and so the linker needs to be told at compile time to ignore errors related to undefined symbols.

This has been handled in this code block:

https://github.com/Chaste/Chaste/blob/baa90ac2819b962188b7562f2326be23c47859a7/heart/CMakeLists.txt#L61-L83

The relevant line for macOS is:
```cmake
target_link_libraries(${dyn_source_name} LINK_PRIVATE "-Wl,-undefined,dynamic_lookup")
```

and this appears to be correct. But the corresponding line for Linux:
```cmake
set_target_properties(${dyn_source_name} PROPERTIES LINK_FLAGS "-Wl,--undefined")
```
does not. The `--undefined` expects an `=<symbol_name>` to indicate which symbol is expected to be undefined.

It appears as though the newer linker available on Ubuntu 24.04 is more picky: presumably, previous versions have been happy to interpret that as "ignore everything".

**Comment by fcooper8472** [comment]
I think the `--undefined` was an attempt to remove the effects of `--no-undefined` that was added as a global linker flag.

The resolution in the attached PR migrates away from using the global link flags (that previously used the `CMAKE_SHARED_LINKER_FLAGS` variable. It was, therefore, impossible to prevent those flags ALSO being applied to the libraries that need to ignore undefined symbols until runtime.
